Refactor TF and TFLite implementations into their own classes/files

reuben opened the pull request:
This refactors the ModelState code which currently into a base `ModelState` class and two implementations, `TFModelState`, `TFLiteModelState`.

This greatly improves the readability of `deepspeech.cc`. I've also taken the opportunity to try and make the code style a bit more consistent. I've made all member variable names end with `_` and fixed inconsistent indentation in a few places.

This also paves the way for fixing the multi-stream bugs we currently have in the client. Not sure if we should take this for 0.5.0 though, as it's a big change.

reuben commented:
Hm, looks like something is wrong with Taskcluster, it didn't pick up the PR.

lissyx commented:
> Hm, looks like something is wrong with Taskcluster, it didn't pick up the PR.

Retriggered them, it looks okay

lissyx commented on native_client/deepspeech.cc:
```diff
@@ -667,182 +276,23 @@ DS_CreateModel(const char* aModelPath,
+  std::unique_ptr<ModelState> model(new TFModelState());
+#else
+  std::unique_ptr<ModelState> model(new TFLiteModelState());
+#endif // USE_TFLITE
```
nit: would be even simpler to read written as:
```
std::unique_ptr<ModelState> model(
#ifndef USE_TFLITE
    new TFModelState()
#else
    new TFLiteModelState()
#endif
);
```

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,258 @@
+              << "file with the same size as the one used for training."
+              << std::endl;
+    return DS_ERR_INVALID_ALPHABET;
+  }
```
nit: maybe this kind of check should be factorized on `ModelState`, like with a `ensure_alphabet(size_t final_dim_size)` that would check against `GetSize()` and return the same error, since the logic is the same for TensorFlow and TFLite runtime, only where we read the value differs ?

reuben replied:
Yeah, I want to do that, ideally we'd uplift all the metadata/model geometry checks to `ModelState`, but that requires finding a good solution for embedding metadata in the TFLite model first.

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,258 @@
+  memcpy(interpreter_->typed_tensor<float>(previous_state_c_idx_), previous_state_c_.get(), sizeof(float) * previous_state_size_);
+  memcpy(interpreter_->typed_tensor<float>(previous_state_h_idx_), previous_state_h_.get(), sizeof(float) * previous_state_size_);
+
+  interpreter_->SetExecutionPlan(acoustic_exec_plan_);
```
nit: I think we should check the return status here.

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,258 @@
+  , previous_state_c_(nullptr)
+  , previous_state_h_(nullptr)
+{
+}
```
No `~TFLiteModelState` ? We're allocating stuff, aren't we?

lissyx replied:
Ok, that makes sense, we `delete` decoder state in `~StreamingState`

reuben commented:
Pushed a commit fixing multiple concurrent streams per model instance.

lissyx commented:
@reuben I guess we can merge that after you rebase, now that 0.5.0 is shipped.

reuben commented:
Re-requesting review for the last commit fixing the concurrent streams bug. I'm also going to add a test that exercises this.

lissyx commented on DeepSpeech.py:
```diff
@@ -618,51 +614,31 @@ def create_inference_graph(batch_size=1, n_steps=16, tflite=False):
+        'input_samples': input_samples,
+    }
+
+    if FLAGS.use_seq_length:
```
I've been thinking about that and I'm wondering, given the reports (myself included) whether we should not also `or not FLAGS.export_tflite` so that passing `--export_tflite` is the only requirement.

lissyx commented on native_client/BUILD:
```diff
@@ -103,34 +114,26 @@ tf_cc_shared_object(
-            "//tensorflow/core/kernels:dense_update_ops",   # Assign
-            "//tensorflow/core/kernels:constant_op",        # Const
-            "//tensorflow/core/kernels:immutable_constant_op", # ImmutableConst
+            "//tensorflow/core/kernels:dense_update_ops",   # Assign (remove once prod model no longer depends on it)
```
Hm, you removed all the `assign` op, are you referring to the prod model we release / host for running tests ?
If so, I do worry that it's a bit burried and we may forgot

reuben replied:
Yep. If we remove this then prod model tests fail when loading the graph.

lissyx replied:
Maybe we should file an issue to not forget and refer to the issue here?

reuben replied:
Filed #2179

lissyx commented on native_client/BUILD:
```diff
@@ -103,34 +114,26 @@ tf_cc_shared_object(
             "//tensorflow/contrib/rnn:lstm_ops_kernels",    # BlockLSTM
-            "//tensorflow/core/kernels:random_ops",         # RandomGammaGrad
             "//tensorflow/core/kernels:pack_op",            # Pack
             "//tensorflow/core/kernels:gather_nd_op",       # GatherNd
```
Have you checked on CUDA runtime ? I remember some of those dependencies were actually only coming from CUDA version.

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,266 @@
+}
+
+void
+TFLiteModelState::copy_vector_to_tensor(const vector<float>& vec,
```
Is it better than just `memcpy` ?

lissyx replied:
Ok, from IRC, this is for zero padding mostly. Might be worth putting a comment.

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,266 @@
+}
+
+void
+TFLiteModelState::copy_tensor_to_vector(int tensor_idx,
```
Same question

reuben replied:
Can't memcpy into an std::vector. This could be implemented as an std::copy call though, if you prefer.

lissyx replied:
I guess any implem is fine, I don't think it's going to be a big bottleneck on the perfs

lissyx commented on native_client/tflitemodelstate.cc:
```diff
@@ -0,0 +1,266 @@
+  copy_vector_to_tensor(mfcc, input_node_idx_, n_frames*mfcc_feats_per_timestep_);
+
+  // Feeding previous_state_c, previous_state_h
+  assert(previous_state_c.size() == state_size_);
```
maybe assert within the copy functions ?

reuben replied:
The copy functions are agnostic to what tensor you're copying stuff into, and they're meant to work with different size vector and `num_elements`, so I can't put these asserts there.

lissyx commented:
@reuben I landed a fix for the ElectronJS build issues, you can rebase on top of it.

reuben commented:
@lissyx I've addressed your comments and also added a test in f12ea5e. The test uses the prod model so it's not actually running, since the graph version was bumped here, but I verified manually. I'll also re-export the 0.5.0 checkpoint and update our prod-model in a separate PR.

reuben commented:
I just saw your messages about 0.5.1 with the DS_IntermediateDecode fix. I'll hold on merging this PR until we do that.
